### PR TITLE
KBC Collision model v2

### PR DESCRIPTION
**Pull request by dominikwilde**
Supporting both AoS and SoA

EnergyReporter added for kinetic energy evaluation

### CODE REVIEW

**Comment by Olllom** [comment]
That looks great!

Could you add a test for the energy reporter mainly to make sure the API works (maybe 2 steps TGV2D on a coarse grid with an interval of 1 and test the initial energy vs. the analytic solution)?

It would also be cool to test some more things for KBC, mainly post-collision shear moments. Maybe you can also find a cheap way to test that gamma optimizes the quadratic entropy approximation? Not required right now, but would be nice.

I will add a few more comments in-line.

File: lettuce/collision.py [hunk 1]
```diff
@@ -37,6 +41,127 @@ def __call__(self, f):
...
+    def __init__(self, lattice, tau):
+        self.lattice = lattice
+        assert lattice.Q == 27, \
+            LettuceException("KBC only realized for D3Q27")
+        self.tau = tau
+        self.beta = 1. / (2 * tau)
+
+        ##Build a matrix that contains the indices
+        self.M = torch.zeros([3, 3, 3, 27], device=lattice.device, dtype=lattice.dtype)
+        if isinstance(lattice,LatticeAoS):
```
Comment (Olllom) [inline lettuce/collision.py]: I am trying to avoid these kinds of distinctions between the AoS and SoA storage. One way would be create M as `lattice.convert_to_tensor(np.zeros(27,3,3,3,dtype=float))`.

Or you can use `lettuce.moments.moment_tensor(lattice.e, torch.tensor([i,j,k]))`

Or we get rid of the LatticeAoS altogether and make our lifes easy.

File: lettuce/io.py [hunk 2]
```diff
@@ -61,3 +61,23 @@ def __call__(self, i, t, f):
...
+        if not isinstance(self.out, list):
+            print("time      kinetic energy", file=self.out)
+
+    def __call__(self, i, t, f):
+        if t % self.interval == 0:
+            u = self.flow.units.convert_velocity_to_pu(self.lattice.u(f))
+            kinE= torch.sqrt(self.lattice.einsum("d,d->", [u,u])[self.lattice.field()])
+            nr_grid_points = self.lattice.rho(f).numel()
+            kinE=kinE.sum()/nr_grid_points
+            print(t, kinE.item())
```
Comment (Olllom) [inline lettuce/io.py]: please use ` file=self.out` and in the error reporter

File: lettuce/io.py [hunk 3]
```diff
@@ -61,3 +61,23 @@ def __call__(self, i, t, f):
...
+        self.interval = interval
+        self.out = [] if out is None else out
+        if not isinstance(self.out, list):
+            print("time      kinetic energy", file=self.out)
+
+    def __call__(self, i, t, f):
+        if t % self.interval == 0:
+            u = self.flow.units.convert_velocity_to_pu(self.lattice.u(f))
+            kinE= torch.sqrt(self.lattice.einsum("d,d->", [u,u])[self.lattice.field()])
+            nr_grid_points = self.lattice.rho(f).numel()
```
Comment (Olllom) [inline lettuce/io.py]: the appropriate thing to do here (instead of dividing by the number of grid points) would be to include the physical length scale --

E = sum ( u_i u_i ) * dx**dim
with `dx = convert_length_to_pu(1.0)`

(Just realized that there is no `convert_length_to_pu` yet but that is trivial to add)

(or you could do the whole calculation in lettuce units and add energy conversion to the units module)

I would also add a method `energy` to the Lattice class - this is a calculation that you will reuse in other situations. Same for things like entropy, etc.  At some point we may port this into a separate module but for now the Lattice class should be the correct place.

Comment (dominikwilde) [reply lettuce/io.py]: Ok, done.

File: lettuce/io.py (hunk 2 above)
Comment (dominikwilde) [reply lettuce/io.py]: Ok

**Comment by Olllom** [comment]
The failure in the CI tests is coming from the `from lettuce import *` statement in the test file.
Some file that is imported has a `from copy import copy` and then this copy is imported.

I would say:
- refactor the `from copy import copy` to import copy in the other file
- leave the `from lettuce import *`

**Comment by Olllom** [comment]
Looks good. The only thing that is missing now is the test for the shear moments (and the local pseudo entropy, now that you have it).

**Comment by Olllom** [comment]
Make sure you get these tests working and this is ready to go.

**Comment by Olllom** [comment]
and resolve the unused variable pointed out by LGTM

File: tests/test_collision.py [hunk 4]
```diff
@@ -36,6 +35,37 @@ def test_collision_fixpoint_2x(Collision, f_all_lattices):
...
+    f, lattice = f_all_lattices
+    tau = 0.5003
+    coll_kbc = Collision(lattice, tau)
+    coll_bgk = BGKCollision(lattice, tau)
+    f_kbc = coll_kbc(f)
+    f_bgk = coll_bgk(f)
+    entropy_kbc = lattice.pseudo_entropy_local(f_kbc)
+    entropy_bgk = lattice.pseudo_entropy_local(f_bgk)
+    assert entropy_bgk.cpu().numpy().max() < entropy_kbc.cpu().numpy().max()
+    assert entropy_bgk.cpu().numpy().min() < entropy_kbc.cpu().numpy().min()
```
[the remark this answers is not in the document]
Comment (dominikwilde) [reply tests/test_collision.py]: This is what I was looking for 👍 Thanks. 

But one all() is enough I guess.


> 
> **To get TRT running on both lattices, you would (a) have to do the `lattice.field` thing in the f indices.**
> But we could also (b) get rid of the `LatticeAoS` first and that should solve it.
> Or, maybe best, we can (c) remove the `LatticeAoS` in the `all_lattices` fixture in `conftest.py` and remove the class alltogether once this PR is through.
> 
> Finally, I have been very sloppy about docstrings so far. But I would still ask you to add some small descriptions of what the functions do (one-liners would do). We can maybe formalize documentation later.

I think we have to refactor quite a lot of code after the removal of AoS, so for the moment I chose (a), which makes it run for the time AoS is still there.

File: lettuce/simulation.py [hunk 5]
```diff
@@ -3,7 +3,7 @@
 from timeit import default_timer as timer
 from lettuce import LettuceException, get_default_moment_transform, BGKInitialization, ExperimentalWarning
 import pickle
-import copy
```
[the remark this answers is not in the document]
Comment (dominikwilde) [reply lettuce/simulation.py]: Won't happen again 🙆

**Comment by Olllom** [comment]
Cool! Thanks so much! Will merge shortly.

**Comment by dominikwilde** [comment]
Remains the question how valuable these extensions are. 

KBC is hard to implement in a pythonic way, so if one day we decide to take it out again due to beauty reasons, then this is also fine. 

TRT is a nice extension I think and in contrast to MRT applicable to any stencil. I tried to reduce the memory consumption as much as possible (and readable).
